packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+      std::unordered_map<lno_t, std::vector<int>> procs_to_send;
+      for(int i = 0; i < exportLIDs.size(); i++){
+        procs_to_send[exportLIDs[i]].push_back(exportPIDs[i]);
+      }
+
+      //call the coloring algorithm
+      twoGhostLayer(nVtx, nVtx+nGhosts, local_adjs_view, offsets, ghost_adjacencies, ghost_offsets,
+                    femv, ownedPlusGhosts, rand, owners2, mapWithCopies, procs_to_send);
+      
+      //copy colors to the output array
+      ArrayRCP<int> colors = solution->getColorsRCP();
+      auto femvdata = femv->getData(0);
+      for(size_t i=0; i<nVtx; i++){
+         colors[i] = femvdata[i];
+      }
+
+    }
+//  private:
+
+    //This is the coloring logic for all 2GL-based methods.
+    //
+    //INPUT ARGS:
+    //  
+    //  n_local: the number of local owned vertices
+    //
+    //  n_total: the number of local owned vertices plus first layer ghosts
+    //
+    //  adjs: the CSR adjacencies for the graph, includes first and second
+    //        layer ghosts
+    //
+    //  offsets: the CSR offsets for the graph, has owned offsets into adjacencies
+    //
+    //  ghost_adjs: the adjacency information for only the second ghost layer
+    //              indexed by ghost_offsets.
+    //              
+    //
+    //  ghost_offsets: the offsets into ghost_adjs, first layer ghost LIDs 
+    //                 minus n_local are used to index this 
+    //                 datastructure. I.E. ghost_offsets(0)
+    //                 corresponds to the ghost with LID n_local
+    //
+    //  gids: a vector that holds GIDs for all vertices on this process
+    //        (includes owned, and all ghosts, indexable by LID)
+    //
+    //  rand: a vector that holds random numbers generated on GID for tie
+    //        breaking. Indexed by LID.
+    //
+    //  owners: contains the process ID for the owner of each vertex.
+    //          Indexable by LID.
+    //
+    //  mapOwnedPlusGhosts: map that can translate between GID and LID
+    //
+    //  procs_to_send: for each vertex that is copied on a remote process,
+    //                 this map contains the list of processes that have
+    //                 a copy of a given vertex. Input LID, get the list
+    //                 of PIDs that have a ghost copy of that LID.
+    //  
+    //OUTPUT ARGS:
+    //  
+    //  femv: an FEMultiVector that holds a dual view of the colors.
+    //        After this call, femv contains updated colors.
+    //
+    void twoGhostLayer(const size_t n_local, const size_t n_total,
+                       const Teuchos::ArrayView<const lno_t>& adjs,
+                       const Teuchos::ArrayView<const offset_t>& offsets,
+                       const Teuchos::ArrayView<const lno_t>& ghost_adjs,
+                       const Teuchos::ArrayView<const offset_t>& ghost_offsets,
+                       const Teuchos::RCP<femv_t>& femv,
+                       const std::vector<gno_t>& gids, 
```

Comment:
Why std::vector when other arguments are ArrayView<const>?

Reply:
Switched them to ArrayViews